src/compiler/commandLineParser.ts

```diff
@@ -884,13 +888,306 @@ namespace ts {
+            switch (valueExpression.kind) {
+                case SyntaxKind.TrueKeyword:
+                    reportInvalidOptionValue(option && option.type !== "boolean");
+                    return true;
+
+                case SyntaxKind.FalseKeyword:
+                    reportInvalidOptionValue(option && option.type !== "boolean");
+                    return false;
+
+                case SyntaxKind.NullKeyword:
+                    reportInvalidOptionValue(!!option);
+                    return null; // tslint:disable-line:no-null-keyword
+
+                case SyntaxKind.StringLiteral:
+                    if (!isDoubleQuotedString(valueExpression)) {
+                        errors.push(createDiagnosticForNodeInSourceFile(sourceFile, valueExpression, Diagnostics.String_literal_with_double_quotes_expected));
+                    }
+                    reportInvalidOptionValue(option && (typeof option.type === "string" && option.type !== "string"));
+                    const text = (<StringLiteral>valueExpression).text;
+                    if (option && typeof option.type !== "string") {
+                        const customOption = <CommandLineOptionOfCustomType>option;
+                        // Validate custom option type
+                        if (!customOption.type.has(text)) {
+                            errors.push(
+                                createDiagnosticForInvalidCustomType(
+                                    customOption,
+                                    (message, arg0, arg1) => createDiagnosticForNodeInSourceFile(sourceFile, valueExpression, message, arg0, arg1)
+                                )
+                            );
+                        }
+                    }
+                    return text;
+
+                case SyntaxKind.NumericLiteral:
+                    reportInvalidOptionValue(option && option.type !== "number");
+                    return Number((<NumericLiteral>valueExpression).text);
+
+                case SyntaxKind.ObjectLiteralExpression:
+                    reportInvalidOptionValue(option && option.type !== "object");
+                    const objectLiteralExpression = <ObjectLiteralExpression>valueExpression;
+
+                    // Currently having element option declaration in the tsconfig with type "object"
+                    // determines if it needs onSetValidOptionKeyValueInParent callback or not
+                    // At moment there are only "compilerOptions", "typeAcquisition" and "typingOptions"
+                    // that satifies it and need it to modify options set in them (for normalizing file paths)
+                    // vs what we set in the json
+                    // If need arises, we can modify this interface and callbacks as needed
+                    if (option) {
+                        const { elementOptions, extraKeyDiagnosticMessage, name: optionName } = <TsConfigOnlyOption>option;

```

Comment:
This cast isn't valid. Someone might have passed an object literal to an option that shouldn't have an object value. We will report an error, but still continue on to this point.